### PR TITLE
Attempt to fix Windows build

### PR DESCRIPTION
**Pull request by rcurtin**
Not an uncommon occurrence, we're seeing C1060 (out of RAM) errors again.  So... the strategy as per usual is to screw around with the options until it works again.  Let's see what happens...

### CODE REVIEW

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

**Comment by rcurtin** [comment]
I can't believe it!  The build succeeded!  I thought the day would never come.  A wrap-up of what I had to do here:

 * Use precompiled headers for all of the mlpack_test build.
 * To do that, the `det_test.cpp` file needed to be adapted to not use the heinous `#define private public` that was being used to test internal members.  I can't believe I did that years ago.
 * Disabled precompiled headers for Linux CI builds because they take up too much RAM/disk and cause build failures on Azure Pipelines.
 * Remove cotire, since that support is now in stock CMake.
 * Fix the ambiguous name `QuadraticFunction`.

File: .ci/macos-steps.yaml [hunk 1]
```diff
@@ -29,7 +28,6 @@ steps:
 
 # Configure mlpack (CMake)
 - script: |
-    unset BOOST_ROOT
```
Comment (shrit) [inline .ci/macos-steps.yaml:32]: still here, after all these years

**Review by shrit — APPROVED** [review]

**Pull request merged**